Evolution/0016-mutli-producer-single-consumer-channel.md

```diff
@@ -0,0 +1,847 @@
+    ///
+    /// Use this source to provide elements to the channel by calling one of the `send` methods.
+    public struct Source: ~Copyable, Sendable {
+        /// A struct representing the backpressure of the channel.
+        public struct BackpressureStrategy: Sendable {
+            /// A backpressure strategy using a high and low watermark to suspend and resume production respectively.
+            ///
+            /// - Parameters:
+            ///   - low: When the number of buffered elements drops below the low watermark, producers will be resumed.
+            ///   - high: When the number of buffered elements rises above the high watermark, producers will be suspended.
+            public static func watermark(low: Int, high: Int) -> BackpressureStrategy
+
+            /// A backpressure strategy using a high and low watermark to suspend and resume production respectively.
+            ///
+            /// - Parameters:
+            ///   - low: When the number of buffered elements drops below the low watermark, producers will be resumed.
+            ///   - high: When the number of buffered elements rises above the high watermark, producers will be suspended.
+            ///   - waterLevelForElement: A closure used to compute the contribution of each buffered element to the current water level.
+            ///
+            /// - Note, `waterLevelForElement` will be called on each element when it is written into the source and when
+            /// it is consumed from the channel, so it is recommended to provide a function that runs in constant time.
+            public static func watermark(low: Int, high: Int, waterLevelForElement: @escaping @Sendable (borrowing Element) -> Int) -> BackpressureStrategy
+
+            /// An unbounded backpressure strategy.
+            ///
+            /// - Important: Only use this strategy if the production of elements is limited through some other mean. Otherwise
+            /// an unbounded backpressure strategy can result in infinite memory usage and cause
+            /// your process to run out of memory.
+            public static func unbounded() -> BackpressureStrategy
+        }
+
+        /// A type that indicates the result of sending elements to the source.
+        public enum SendResult: ~Copyable, Sendable {
+            /// An opaque token that is returned when the channel's backpressure strategy indicated that production should
+            /// be suspended. Use this token to enqueue a callback by  calling the ``MultiProducerSingleConsumerAsyncChannel/Source/enqueueCallback(callbackToken:onProduceMore:)`` method.
+            ///
+            /// - Important: This token must only be passed once to ``MultiProducerSingleConsumerAsyncChannel/Source/enqueueCallback(callbackToken:onProduceMore:)``
+            ///  and ``MultiProducerSingleConsumerAsyncChannel/Source/cancelCallback(callbackToken:)``.
+            public struct CallbackToken: Sendable, Hashable { }
```

Comment:
Is it possible that these tokens are somehow user provided?

Reply:
If you mean constructed by the user then no. There is no public `init` for those.

Reply:
no, I was suggesting that the user provides a type to represent a token and this could be token that the framework provides. It would mean that things would need to be flipped around to be generic around that.

There are some APIs that have tokens associated with callbacks. Thinking particularly the CF APIs or the KVO APIs

Reply:
I see what you mean I would like to avoid that though since we are using those tokens to understand if there is a potential API mis-use. We wouldn't be able to do this if the user were to create those tokens.